### PR TITLE
feat: add types for Typescript usage

### PR DESCRIPTION
**Pull request by ormuscama**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — CHANGES_REQUESTED** [review]
Thanks for the contribution! Could you add a test using `tsd`? See all other plugins in this org.

**Comment by ormuscama** [comment]
Yes of course.
I will also move index.d.ts at the same level as index.ts to be closer to what is done in other plugins

File: test/types/index.test-d.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { FastifyInstance, LightMyRequestResponse } from 'fastify'
+import { expectAssignable, expectType, expectError } from 'tsd'
+import { FastifyRestartableOptions, start } from '../..'
+
```
Comment (mcollina) [inline test/types/index.test-d.tsx:4]: Why `.tsx`? Just use `.ts`.

Comment (ormuscama) [reply test/types/index.test-d.tsx]: typo ( was doing some react ...)

File: package.json [hunk 2]
```diff
@@ -4,8 +4,10 @@
   "description": "Restart Fastify without losing a request",
   "main": "index.js",
   "scripts": {
-    "test": "standard | snazzy && tap test.js"
+    "test": "standard | snazzy && tap test/test.js",
+    "test:tsd": "tsd"
```
Comment (mcollina) [inline package.json:8]:
```suggestion
    "test": "standard | snazzy && tap test/test.js && tsd"
```

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**